src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -661,22 +662,22 @@ internal static T[] AllocateUninitializedArray<T>(int length)
             }
 
             if (length < 0)
+            {
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.lengths, 0, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
-#if DEBUG
-            // in DEBUG arrays of any length can be created uninitialized
-#else
-            // otherwise small arrays are allocated using `new[]` as that is generally faster.
-            //
-            // The threshold was derived from various simulations.
-            // As it turned out the threshold depends on overal pattern of all allocations and is typically in 200-300 byte range.
-            // The gradient around the number is shallow (there is no perf cliff) and the exact value of the threshold does not matter a lot.
-            // So it is 256 bytes including array header.
-            if (Unsafe.SizeOf<T>() * length < 256 - 3 * IntPtr.Size)
+            }
+
+#if !DEBUG  // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
+            // small arrays are allocated using `new[]` as that is generally faster.
+            if (length < 2048 / Unsafe.SizeOf<T>())
             {
                 return new T[length];
             }
 #endif
-            return (T[])AllocateNewArray(typeof(T[]).TypeHandle.Value, length, zeroingOptional: true);
+            // kept outside of the small arrays hot path to have inlining without big size growth
+            return AllocateNewUninitializedArray(length);
+
+            T[] AllocateNewUninitializedArray(int length)
```

Comment:
Maybe add a comment here with the link to the issue about intrinsic `TypeHandle.Value` so that this split can be undone once that issue is fixed?

Reply:
good idea, done